### PR TITLE
Add support for 2023 16/18 Blade models.

### LINKED ISSUE #2115: Support for Blade 18
### Device Info

* Name: Blade 18
* Product Number: RZ09-0484
* Model Number:  RZ09-0484 
* (not sure if these numbers are supposed to be the same, but in the booklet that's given as Model Nr and on the underside of the laptop this is listed as the Product Nr.)

### Device Information
```
$  lsusb -d 1532: -v                                                                                                                                                                    

Bus 001 Device 003: ID 1532:02a0 Razer USA, Ltd Razer Blade
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x02a0
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Blade
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      2 Mouse
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1

```

```
$ ls -lah /dev/input/by-id/                                                                                                                                                             
total 0
drwxr-xr-x 2 root root 140 Aug 16 15:44 .
drwxr-xr-x 4 root root 620 Aug 16 15:44 ..
lrwxrwxrwx 1 root root   9 Aug 16 15:44 usb-Razer_Razer_Blade-event-if01 -> ../event6
lrwxrwxrwx 1 root root   9 Aug 16 15:44 usb-Razer_Razer_Blade-event-kbd -> ../event4
lrwxrwxrwx 1 root root   9 Aug 16 15:44 usb-Razer_Razer_Blade-if01-event-kbd -> ../event5
lrwxrwxrwx 1 root root   9 Aug 16 15:44 usb-Razer_Razer_Blade-if02-event-mouse -> ../event7
lrwxrwxrwx 1 root root   9 Aug 16 15:44 usb-Razer_Razer_Blade-if02-mouse -> ../mouse0

```

<details>

```

#   usbhid-dump -m 1532 -ed
001:003:002:DESCRIPTOR         1692194574.783386
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 09 40 75 08 95 02 15 81 25 7F 81 02 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 06 00
 FF 09 02 15 00 25 01 75 08 95 5A B1 01 C0

001:003:001:DESCRIPTOR         1692194574.784134
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 0E 81 00 C0 05 0C 09 01 A1 01 85
 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75 10 81
 00 75 08 95 0D 81 01 C0 05 01 09 80 A1 01 85 03
 19 81 29 83 15 00 25 01 75 01 95 03 81 02 95 05
 81 01 75 08 95 0E 81 01 C0 05 01 09 00 A1 01 85
 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75 08 95
 0F 81 00 C0 05 01 09 00 A1 01 85 05 09 03 15 00
 26 FF 00 35 00 46 FF 00 75 08 95 0F 81 00 C0

001:003:000:DESCRIPTOR         1692194574.784801
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

```

</details>

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on **Razer Synapse** in a Windows Environment>

Will look into creating packet captures and screenshots later.

I think this should be pretty similar to #2092, iirc they have basically the same keyboard. I will hopefully have some time soon to see if a simple translation of the PR to the Blade 18 does anything useful.

### PR DESCRIPTION
**Pull request by nanomatters**
Fixes #2115

### CODE REVIEW

**Comment by nanomatters** [comment]
image

File: .gitignore [hunk 1]
```diff
@@ -65,3 +65,6 @@ dist
 /driver/*.mod
 /driver/*.mod.c
 /driver/.*.cmd
+
+# Visual Studio Code files
+.vscode
```
Comment (whot) [inline .gitignore]: not a project member but this should usually be split out into a separate commit

Comment (nanomatters) [reply .gitignore]: I lost count of projects and the companies that I worked for using git and I have never in my life have seen a project that asked me to deliver the .gitignore file in a separate commit.

Comment (z3ntu) [reply .gitignore]: Agreed with @whot , that should be a separate commit since it has nothing to do with the device support. 

I'd even say since you are adding two separate devices those should be separate commits, but no worries about that.

File: README.md [hunk 2]
```diff
@@ -97,6 +97,8 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer DeathStalker V2 Pro (Wired)             |  1532:0292  |
 | Razer DeathStalker V2                         |  1532:0295  |
 | Razer BlackWidow V3 Tenkeyless                |  1532:0A24  |
+| Razer Blade 16 (2023)                         |  1532:029F  |
+| Razer Blade 18 (2023)                         |  1532:02A0  |
```
Comment (z3ntu) [inline README.md:105]: Please keep those entries sorted by PID, so move them up to where they belong

File: driver/razerkbd_driver.h [hunk 3]
```diff
@@ -94,6 +94,8 @@
 #define USB_DEVICE_ID_RAZER_DEATHSTALKER_V2_PRO_WIRED 0x0292
 #define USB_DEVICE_ID_RAZER_DEATHSTALKER_V2 0x0295
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_TK 0x0A24
+#define USB_DEVICE_ID_RAZER_BLADE_16_2023 0x029F
+#define USB_DEVICE_ID_RAZER_BLADE_18_2023 0x02A0
```
Comment (z3ntu) [inline driver/razerkbd_driver.h:102]: keep these also sorted by value

File: install_files/udev/99-razer.rules [hunk 4]
```diff
@@ -10,7 +10,7 @@ ATTRS{idProduct}=="0013|0016|0020|0024|0025|0029|002e|002f|0032|0034|0036|0037|0
     ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
 
 # Keyboards
-ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|024e|0252|0253|0255|0256|0257|0258|0259|025a|025d|025e|0266|0269|026a|026b|026c|026d|026e|026f|0270|0271|0276|0279|027a|0282|028a|028b|028c|0290|0292|0295|0a24", \
+ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|024e|0252|0253|0255|0256|0257|0258|0259|025a|025d|025e|0266|0269|026a|026b|026c|026d|026e|026f|0270|0271|0276|0279|027a|0282|028a|028b|028c|0290|0292|0295|0a24|02a0|029f", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules]: same, keep them sorted please

**Comment by z3ntu** [comment]
Just some small style issues, otherwise diff looks good.

Just to confirm, you have tested this on both devices and all the added functionality works correctly?

**Comment by youssefKadaouiAbbassi** [comment]
hello, just wanted to ask if the support of razer balde 16 and 18 will be available soon

**Comment by z3ntu** [comment]
@youssefKadaouiAbbassi When this pull request is ready (all issues resolved and working well), I'll merge it and then it'll be included in the next release.

Until then you could build it yourself from source: https://github.com/openrazer/openrazer/wiki/Building-a-package

**Comment by vtretyakov** [comment]
@nanomatters thanks for making this effort to support the newest models. I tried your branch on Ubuntu 20.04 with the Razer Blade 16 (2023) and have no difference after the openrazer installation. The touchpad, sound and microphone do not work. Could you help me to figure out how to solve it?

**Comment by z3ntu** [comment]
@vtretyakov 

> The touchpad, sound and microphone do not work.

Nothing is expected to change with openrazer support. With the support here you'll be able to change RGB lights etc but generally not core functionality.

@nanomatters Still interested in fixing the comments I've left on the diff? If they are fixed I can merge this, assuming the functionality works well.

**Comment by imLew** [comment]
@z3ntu I think your comments were addressed in 266ee44, the comments just weren't marked as resolved.

**Comment by z3ntu** [comment]
@imLew Right, somehow missed that.

If you could build openrazer from source from this PR and validate this works, I'd be grateful! https://github.com/openrazer/openrazer/wiki/Building-a-package (Also let me know which of the two devices you have)

**Comment by imLew** [comment]
I just built the package from @nanomatters fork on the Blade 18 and razergenie and razercommander mostly work.
All the different pre-configured lighting effects for the keyboard work.
In razergenie turning the logo on and off works (I didn't find this option in razercommander), however it also shows options for 'Ripple' and 'Ripple Random', which actually toggle the ripple effect on the keyboard. 

If there is anything else I can test let me know. I will be probably trying some other tools tomorrow.

**Comment by z3ntu** [comment]
> In razergenie turning the logo on and off works (I didn't find this option in razercommander)

Good :) Btw razercommander isn't maintained since many years so it's likely it doesn't support many options on newer devices.

> however it also shows options for 'Ripple' and 'Ripple Random', which actually toggle the ripple effect on the keyboard.

So this also works? Sounds good then.

**Comment by imLew** [comment]
> Good :) Btw razercommander isn't maintained since many years so it's likely it doesn't support many options on newer devices.

Thanks, it's my first razer so I just installed a bunch of stuff I found in the AUR for `razer`.

> >  however it also shows options for 'Ripple' and 'Ripple Random', which actually toggle the ripple effect on the keyboard.
> 
> So this also works? Sounds good then.
No, what I mean is, there is a dropdown menu for the logo effect that includes 'Ripple', but choosing this option switches to the ripple effect for the keyboard and does not affect the logo. 

ETA: I will merge current master tomorrow and try again to be sure.
Anything else I should check?

**Comment by z3ntu** [comment]
I'd suggest polychromatic also, it's a really nice UI. But generally test procedure is clicking around in the UI, seeing if anything obvious is missing there and all the options in the UI should work on the device mostly reliably.

**Comment by imLew** [comment]
Thanks for the recommendation, it looks very sleek.

Anyway, I created my own fork to merge master into it and built from there. The instruction for building from a local repo didn't work for me.

I'm not familiar with building/installing kernel modules at all, so after building the package I ran `dkm install` with the `--force` option, I hope that results in using the updated build.

Either way, everything in polychromatic worked :D

**Comment by z3ntu** [comment]
@imLew Btw which device do you have? `lsusb | grep 1532` should show that

**Comment by imLew** [comment]
A blade 18
`Bus 001 Device 003: ID 1532:02a0 Razer USA, Ltd Razer Blade`

**Review by z3ntu — APPROVED** [review]
LGTM!

@imLew Thanks for testing. I'm going to assume that the other model added here (0x029F) also works even though I didn't get direct confirmation on it by someone who has that device.
